app/src/components/main.tsx

```diff
@@ -19,31 +19,37 @@ import { AppSelectScript } from './select-script'
 import {
   pythonReady, pythonData, IPythonData, pythonCode
 } from '../observables/python'
-import { inputDirectory, outputDirectory } from '../observables/directories'
+import { inputDirectory, outputDirectory } from '../observables/directories';
+import {
+  sendExecuteRequest, sendFileTransferRequest, sendFileTransfer
+} from '../observables/webworker-messaging/main';
 
-import runUserCode from '../python/run-user-code.py';
 import zipOutput from '../python/zip-output.py';
 import updateOutput from '../python/update-output.py';
 
 
-declare let pyodide: any;
-declare var Module: any;
-
-
 function runConversion() {
-  pyodide.runPython(runUserCode)
-    .catch(() => {
-      pyodide.runPython(updateOutput)
-    })
-    .then(() => {
-      pyodide.runPython(updateOutput)
+  const code = pythonCode.getValue();
+  // const code = "print('boo'); open('/output/test', 'a').close()"
+  sendExecuteRequest(code).subscribe(() => {
+    sendExecuteRequest(updateOutput).subscribe(result => {
+      const fileNames = result.data.result
+      outputDirectory.next(fileNames)
     })
+  })
 }
 
+
 function downloadOutput() {
-  pyodide.runPython(zipOutput).then(() => {
-    let zip = Module.FS.readFile('/output.zip') as Uint8Array
-    saveAs(new Blob([new Uint8Array(zip)]), 'output.zip')
+  const filename = '/output.zip'
+
+  sendExecuteRequest(zipOutput).subscribe(() => {
```

Comment:
Created a new function called `sendExecuteRequest` this is designed so that it can work whether or not pyodide is within a web worker.